core/dive.c

```diff
@@ -492,7 +492,9 @@ static void dc_cylinder_renumber(struct dive *dive, struct divecomputer *dc, con
 /* copy dive computer and renumber the cylinders */
 static void copy_dc_renumber(struct dive *d, const struct divecomputer *sdc, struct divecomputer *ddc, const int cylinders_map[])
 {
+	struct divecomputer *old_next = ddc->next;
 	*ddc = *sdc;
+	ddc->next = old_next;
```

Comment:
So that's rather scary. I don't think that's what we should do. Because now the _second_ and all further DCs are those of the original dive.

Here's how that C data structure beauty works...
struct dive CONTAINS a struct divecomputer - so the first dive computer is always part of the dive structure.
And any **additional** dive computer is then attached via the linked list.
So if you copy the `->next` pointer, you basically attach whatever second and up dive computer of ddc were before this function was called to the new dive. I need to read this a couple more times in context, but that does feel wrong.

Reply:
Yes I understand the data structure. :)
I was hoping that copy_dc_renumber was used to extend the divecomputer-list of the newly generated destination dive. So shouldn't ddc->next always be NULL (because freshly generated)?

If this:
```
        if (a->model && !b->model) {
                copy_dc_renumber(d, a, res, cylinders_map_a);
                return;
        }
```
is supposed to copy all dive computers, then clearly the dc_renumber has to be put in a loop, isn't it....?

Reply:
Sorry, didn't intend to sound condescending.
I don't think we can assume that ddc->next is always NULL because we could be importing (instead of downloading) and an import could contain more than one DC per dive (which isn't the case in my example, though).

Reply:
I see something else that's scary, though.

`dive.c:3445 	const struct dive *dl = NULL;`
next use is
`dive.c: 3487  		join_dive_computers(res, &res->dc, &dl->dc, &a->dc, cylinders_map_b, cylinders_map_a, 1);`
Umm, that does seem like a bad idea...

Reply:
> Sorry, didn't intend to sound condescending.

You didn't.

> I don't think we can assume that ddc->next is always NULL

It should be, because ddc belongs to a freshly constructed dive.

But you are correct - the code is wrong. This all needs to be put in loops to copy the whole list.

Reply:
oh, and that was introduced in your 8c2383b495

Reply:
> I see something else that's scary, though.

Good catch - that should by `&b->dc`.

Reply:
> I get the feeling that you might need to copy the whole chain.

Yes, this is definitely the case. I tried to implement something, but I am not sure I did this correctly. Please check if you find time.

>  And if you feel like you want to rewrite dive.c as C++, I think that's ok. 

Hell no. For many reasons, one of them that I don't even know a small fraction of the use cases (CCR, multi-cylinder, multi-dive computer, deco, ...).

Reply:
> > And if you feel like you want to rewrite dive.c as C++, I think that's ok.
> 
> Hell no.

That made me laugh :-)